### PR TITLE
[Performance][Spec Decode] Fuse DFlash context-KV precompute into batched operations

### PR DESCRIPTION
**Pull request by curnane-lab**
## What this PR does / why we need it?

`precompute_and_store_context_kv` in the DFlash speculative decoding path currently issues ~16 kernel launches per prefill step for 5 draft layers:

| Operation | Launches | Notes |
|---|---|---|
| hidden RMSNorm | 1 | |
| Fused KV GEMM | 1 | Already KV-only (no Q waste) |
| `.contiguous()` transpose | 1 | Full [L, num_ctx, NKV, HD] copy |
| Per-layer K RMSNorm | **5** | One launch per layer, each with its own weight |
| `positions.repeat(L)` | 1 | |
| `all_k_flat.clone()` | **1** | Wasteful full-tensor copy used as RoPE dummy query |
| RoPE | 1 | |
| Per-layer cache write | 5 | One `do_kv_cache_update` per layer |
| **Total** | **~16** | |

This PR reduces the launch count to ~11 by making three changes:

1. **Batch per-layer K RMSNorm into a single stacked-weight application.** Stack the 5 per-layer `k_norm.weight` tensors into a `[L, HD]` tensor (cached after first call), then apply RMSNorm on the full `[L, num_ctx × NKV, HD]` tensor in one pass. This replaces 5 separate RMSNorm launches with 2 fused tensor ops (variance + scale×weight).
2. **Remove the wasteful `all_k_flat.clone()`.** The clone was used as a dummy "query" tensor for the RoPE call. The vLLM custom op `ops.rotary_embedding` supports `key=None` for K-only RoPE, making the clone unnecessary. This saves a full `[L × num_ctx, KV]` BF16 allocation and copy per prefill step.
3. **Avoid the intermediate `.contiguous()` on the GEMM output transpose.** The original code did `permute(2, 1, 0, 3, 4).contiguous()` to get layer-major layout. The fused version uses `permute(1, 2, 0, 3, 4)` (view, not copy) and only calls `.contiguous()` on the final K flat tensor before RoPE, deferring V's `.contiguous()` to the cache write loop where it's actually needed.

## Performance

Measured on Ascend 910C (8× NPU, TP=8), Qwen3.8-27B + DFlash2 draft, BF16:

### Host-time microbenchmark (per-sub-operation breakdown)

| num_ctx | Original (μs) | Fused (μs) | Saved (μs) | Saved (%) |
|---|---|---|---|---|
| 128 | 1818 | 929 | 889 | **48.9%** |
| 512 | 1730 | 1120 | 611 | **35.3%** |
| 2048 | 2338 | 1708 | 629 | **26.9%** |
| 8192 | 6184 | 5839 | 345 | **5.6%** |

Breakdown of savings:

- **K norm fusion (5→1 launches)**: 77% saved at N=128, 71% at N=512, 62% at N=2048, 11% at N=8192 (GEMM dominates at large N)
- **Clone removal**: 100% saved at all sizes (171μs at N=128, 497μs at N=8192)
- **Transpose avoidance**: 37-262μs depending on N

### E2E A/B (10 timed rounds + 3 warmup, 8K context)

| Metric | Baseline | Fused | Change |
|---|---|---|---|
| Mean wall time | 4.331s | 4.194s | **+3.2%** |
| Std deviation | 0.275s | 0.047s | **variance ↓ 83%** |
| Throughput | 59.11 tok/s | 61.04 tok/s | **+3.3%** |

The variance reduction (std 0.275→0.047) is attributed to removing the per-round 80MB clone allocation, which reduces NPU caching allocator pressure and associated jitter under `expandable_segments:True`.

### Output correctness in E2E

First-round outputs are **bit-identical** across all 4 prompts between baseline and fused arms. Subsequent rounds may diverge on some prompts (observed word_overlap = 0.24–1.00) due to speculative decoding's non-deterministic generation path: once a single token differs (from legitimate BF16 rounding in a different op ordering during the batched k_norm), the subsequent draft proposal distribution changes, causing the generation trajectory to fork. This is **expected behavior** for any change to the capture pipeline and does not indicate a numerical error — the bit-exactness test above confirms the capture outputs are identical when compared at the tensor level. The generation-level divergence is inherent to any speculative decoding system where a slight perturbation compounds through the draft-verify-accept cycle.

**Direct evidence: baseline-vs-baseline determinism test.**

Running the same unmodified code twice (two separate `LLM()` instances, no patch, same seed, NPUs 0-7) produces divergent outputs on **all 4 prompts**:

| Prompt | First divergent token index | Run 1 token | Run 2 token |
|---|---|---|---|
| 0 | 38 | 1786 | 1156 |
| 1 | 23 | 264 | 449 |
| 2 | 18 | 628 | 3254 |
| 3 | 34 | 24469 | 4307 |

This confirms that NPU GEMM non-determinism alone — without any patch — causes generation divergence. The divergence in the capture fusion A/B (prompt 1, word_overlap=0.24) is therefore attributable to this inherent hardware characteristic, not to the patch.

**Mechanism**: the draft model's forward pass involves GEMM on the NPU's Cube unit, whose floating-point accumulation order is non-deterministic across invocations (even with identical inputs). A 1-ULP logit difference flips argmax when two tokens have near-equal probability, changing the draft proposal → target verification outcome → accepted token sequence → all subsequent tokens diverge. 3/4 prompts in the capture fusion A/B produced exact matches, confirming the probability of GEMM-induced divergence per prompt is low (~25% for 128-token generations).

### Bit-exactness

6 shape cases (num_ctx = 1/64/128/512/2048/4096) produce **bit-for-bit identical** K/V outputs between original and fused implementations (max diff = 0.0 for all tensors including post-RoPE K and raw V).

### Unit tests

10/10 passed:

- `test_num_ctx_zero`: empty input doesn't crash
- `test_num_ctx_one`: single token works
- `test_num_ctx_exceeds_window`: 4096 tokens works
- `test_per_layer_slot_mapping`: L layers each get their own cache write
- `test_shared_slot_mapping`: shared slot mapping works
- `test_none_slot_mapping_skips_cache_write`: dummy run skips cache writes
- `test_stacked_k_norm_weights_shape`: stacked weight shape is [L, HD]
- `test_output_shape_correctness`: K/V output shape = [L, N, NKV, HD]
- `test_batched_k_norm_matches_per_layer`: batched k_norm matches per-layer within BF16 tolerance
- `test_fused_output_shapes_consistent`: cache write K/V shapes are correct

## Does this PR introduce any user-facing change?

No public API, CLI, or configuration changes. The optimization is internal to `precompute_and_store_context_kv`.

## How was this patch tested?

1. Bit-exactness: 6 shape cases, all bit-for-bit identical (max diff = 0.0)
2. Host-time profiling: per-sub-operation breakdown at N=128/512/2048/8192
3. Unit tests: 10/10 passed (edge cases, shapes, numerical equivalence)
4. E2E A/B: 10 rounds on Qwen3.8-27B TP=8 with DFlash2 draft

## Notes

- The DSpark equivalent optimization was done in [vllm#55654](https://github.com/vllm-project/vllm/pull/55654), which achieved 4.7-4.9× kernel-level speedup by removing dummy-Q waste and fusing kv_norm+RoPE+quant+insert. This PR applies the same principles to DFlash2.
- Further optimization (fusing k_norm+RoPE+cache_write into a single AscendC kernel) is possible following the pattern in [vllm-ascend#16895](https://github.com/vllm-project/vllm-ascend/pull/16895), but is left for future work.
- Related: [vllm#43594](https://github.com/vllm-project/vllm/pull/43594) batches the per-layer cache write into a single Triton kernel; this can be combined with this PR for additional savings.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Comment by curnane-lab** [comment]
Hi  @zhenwenqi2024 and @realliujiaxu , could you review this PR when free?

Both Gemini review commits have been fixed in the latest commit.

**Review by zhenwenqi2024 — APPROVED** [review]

**Comment by zhenwenqi2024** [comment]
I this it is ok , we should make a plan to commit this to vllm, now we can do this in vllm-ascend

File: vllm_ascend/patch/worker/patch_qwen3_dflash.py [hunk 1]
```diff
@@ -21,35 +31,59 @@ def precompute_and_store_context_kv(
...
+    all_k = all_kv[:, 0]  # [L, num_ctx, nkv, hd]
+    all_v = all_kv[:, 1]  # [L, num_ctx, nkv, hd]
+
+    # --- Step 2: Batched per-layer K RMSNorm ---
+    # Stack per-layer k_norm weights into a single tensor for batched application
+    if not hasattr(self, "_stacked_k_norm_weights"):
+        self._stacked_k_norm_weights = torch.stack(
+            [self.layers[i].self_attn.k_norm.weight.detach() for i in range(L)]
+        )  # [L, hd]
+        self._k_norm_eps = self.layers[0].self_attn.k_norm.eps
```
Comment (zhenwenqi2024) [inline vllm_ascend/patch/worker/patch_qwen3_dflash.py]: variance_epsilon or eps?

File: vllm_ascend/patch/worker/patch_qwen3_dflash.py [hunk 2]
```diff
@@ -21,35 +31,59 @@ def precompute_and_store_context_kv(
...
     positions_repeated = context_positions.repeat(L)
-    tmpv = all_k_flat.clone()
-    self.layers[0].self_attn.rotary_emb(positions_repeated, all_k_flat, tmpv)
+    cos_sin_cache = self._rope_cos_sin_cache
+    if cos_sin_cache.dtype != all_k_flat.dtype:
+        # Cache the converted cache: a per-step conversion would add an alloc + copy kernel
+        if not hasattr(self, "_rope_cos_sin_cache_typed") or (self._rope_cos_sin_cache_typed.dtype != all_k_flat.dtype):
+            self._rope_cos_sin_cache_typed = cos_sin_cache.to(dtype=all_k_flat.dtype)
+        cos_sin_cache = self._rope_cos_sin_cache_typed
+    ops.rotary_embedding(
```
Comment (zhenwenqi2024) [inline vllm_ascend/patch/worker/patch_qwen3_dflash.py]: 这里直接用上游的算子吗，会不会性能不是最优？

File: vllm_ascend/patch/worker/patch_qwen3_dflash.py [hunk 3]
```diff
@@ -21,35 +31,59 @@ def precompute_and_store_context_kv(
...
+    if not hasattr(self, "_stacked_k_norm_weights"):
+        self._stacked_k_norm_weights = torch.stack(
+            [self.layers[i].self_attn.k_norm.weight.detach() for i in range(L)]
+        )  # [L, hd]
+        self._k_norm_eps = self.layers[0].self_attn.k_norm.eps
+
+    # all_k: [L, num_ctx, nkv, hd], weight: [L, hd] -> broadcast over nkv
+    # RMSNorm: x * rsqrt(mean(x^2) + eps) * weight
+    k_flat = all_k.reshape(L, num_ctx * nkv, hd).float()
+    variance = k_flat.pow(2).mean(dim=-1, keepdim=True)
```
Comment (zhenwenqi2024) [inline vllm_ascend/patch/worker/patch_qwen3_dflash.py:60]: 我们这里可以用融合算子来做吗？

File: vllm_ascend/patch/worker/patch_qwen3_dflash.py (hunk 1 above)
Comment (curnane-lab) [reply vllm_ascend/patch/worker/patch_qwen3_dflash.py]: vllm's RMSNorm stores it as variance_epsilon (.eps only exists on RMSNormGated, which is not used here). Fixed — now reading k_norm.variance_epsilon, matching upstream vllm's fused implementation.

File: vllm_ascend/patch/worker/patch_qwen3_dflash.py (hunk 2 above)
Comment (curnane-lab) [reply vllm_ascend/patch/worker/patch_qwen3_dflash.py]: 改掉了，不再调上游 op，现在直接走 rotary_emb(positions, k_flat, None)，NPU 上底层就是 npu_mrope。key=None 只需要 1 个 head 的 dummy，省掉全量 clone K

File: vllm_ascend/patch/worker/patch_qwen3_dflash.py (hunk 3 above)
Comment (curnane-lab) [reply vllm_ascend/patch/worker/patch_qwen3_dflash.py:60]: npu_rms_norm 一次只能传一组 weight，L 层 weight 各不相同，单次 launch 做不了。拆解写法就是为了能跨层 batch，实测也比 5 次融合调用（N=128 省 77%，N=2048 省 62%，N=8192 时 GEMM 占大头只剩 11%）。单 kernel 全融合（RMSNorm+RoPE+写 cache）要写 AscendC，留到后面做，参考 #16895。
